shared/src/balancer/logexpmath.rs

```diff
@@ -0,0 +1,520 @@
+//! Module emulating the pow function as found in the Balancer contract code.
+// The original contract code can be found at:
+// https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/solidity-utils/contracts/math/LogExpMath.sol
+
+#![allow(dead_code)]
+
+use ethcontract::{I256, U256};
+use lazy_static::lazy_static;
+use std::convert::TryFrom;
+
+lazy_static! {
+    static ref ONE_18: I256 = I256::from(10_u128.pow(18));
+    static ref ONE_20: I256 = I256::from(10_u128.pow(20));
+    static ref ONE_36: I256 = I256::from(10_u128.pow(36));
+    static ref UNSIGNED_ONE_18: U256 = U256::try_from(*ONE_18).unwrap();
+    static ref MAX_NATURAL_EXPONENT: I256 = ONE_18.checked_mul(I256::from(130_i128)).unwrap();
+    static ref MIN_NATURAL_EXPONENT: I256 = ONE_18.checked_mul(I256::from(-41_i128)).unwrap();
+    static ref LN_36_LOWER_BOUND: I256 = ONE_18.checked_sub(I256::from(10_u128.pow(17))).unwrap();
+    static ref LN_36_UPPER_BOUND: I256 = ONE_18.checked_add(I256::from(10_u128.pow(17))).unwrap();
+    static ref MILD_EXPONENT_BOUND: I256 =
+        I256::from(2_u32).pow(254_u32).checked_div(*ONE_20).unwrap();
+}
+fn constant_x_20(i: u32) -> I256 {
+    match i {
+        2 => 3_200_000_000_000_000_000_000_i128,
```

Comment:
I see that the balancer code has x0 to x11 and a0 to a11 as constants. Maybe it would be more ergonomic to store them in a fixed size array here? like `const X_: [I256; 11] = [...]`. I'm not sure, just an idea.

Also as the balancer constants have an explanation for how they were calculated like
>  int256 constant x0 = 128000000000000000000; // 2ˆ7

We could define them as `I256::exp10(18 + 7)`. On the other hand I can see the argument for keeping this code as close to the balancer code as possible.

Reply:
I could definitely use only two arrays for `x` and `a`. I decided to split each in two arrays (`18` and `20`) because the first two of the array are scaled compared to the other 9 by a factor of 100. This in particular means that they are used in different places in the code, and if an index is used in the wrong place then the function panics. This was helpful when writing up everything to catch off-by-one issues. On the other end, I'm also ok with using an array and being more consistent with the original Balancer code.

About computing the constants: at first I did that for the `x` constants, but if we want to stay consistent and do the same for the `a` constants I'd also need to truncate the decimal representation of `e^n` after some point, which looks overly complicated for me to this purpose. So I also kept `x` as a list for consistency. Again no strong opinion, thanks to the tests we could define everything with an explicit calculation without any issue.